src/transformers/models/llama/modeling_llama.py

```diff
@@ -1070,9 +1051,12 @@ def _update_causal_mask(
                 return attention_mask
             return None
 
-        if self.config._attn_implementation == "sdpa":
-            # For SDPA, when possible, we will rely on its `is_causal` argument instead of its `attn_mask` argument,
-            # in order to dispatch on Flash Attention 2.
+        # For SDPA, when possible, we will rely on its `is_causal` argument instead of its `attn_mask` argument, in
+        # order to dispatch on Flash Attention 2. This feature is not compatible with static cache, as SDPA will fail
+        # to infer the attention mask.
+        past_seen_tokens = past_key_values.get_seq_length() if past_key_values is not None else 0
+        using_static_cache = isinstance(past_key_values, StaticCache)
```

Comment:
As I understand it, once the StaticCache is initialized, there is no need to pass it in `past_key_values` argument. That's why additional condition is necessary. Suggestion:
`using_static_cache = isinstance(past_key_values, StaticCache) or isinstance(
            getattr(self.layers[0].self_attn, "past_key_value", None), StaticCache
        )`

Reply:
@poedator This PR changes precisely the assumption you wrote: we will always need to pass the cache, after this PR it is an object that does NOT live inside the model.

This change will make the transformers' team work easier 🤗